### PR TITLE
feat: Add PatchCollection

### LINKED ISSUE #2389: Add `PatchCollection` function to `Store`
Like `PatchSchema`, but for `CollectionDescription`s.

This is a prerequisite for moving more stuff from schema onto collection, as users will need a means to set the moved properties.

Should eventually allow the setting of indexes and ACP stuff too, but it is fine to block those off initially in order to reduce scope of the initial PR.

As the only to-be mutable property that is not index/acp and currently already on collection description is col/field name, and those properties comes with higher risk, it might be best to do this alongside https://github.com/sourcenetwork/defradb/issues/1386 or https://github.com/sourcenetwork/defradb/issues/1385 to keep things simple to start off with.

### PR DESCRIPTION
**Pull request by AndrewSisley**
## Relevant issue(s)

Resolves #2389

## Description

Adds the PatchCollection command.

Mutating anything but the collection name is currently disabled, we can expand this as we see fit, but for now I'd prefer to keep the initial PR small.

This change means that the Collection Name is no longer always going to be the same as the Schema Name.

I've manually tested the OpenApi stuff via playground.

### CODE REVIEW

File: cli/collection_patch.go
Comment (shahzadlone) [inline cli/collection_patch.go:1]: todo: 
```suggestion
// Copyright 2024 Democratized Data Foundation
```

File: cli/collection_patch.go
Comment (shahzadlone) [inline cli/collection_patch.go:58]: question: In what case can `len(args)` > 1?

File: client/db.go
Comment (shahzadlone) [inline client/db.go:128]: question: what is meant by `valid state` here.
```suggestion
	// individual operations defined in the patch do not need to result in a `valid state`, only the net result
```

File: client/db.go
Comment (shahzadlone) [inline client/db.go:127]: todo: This sentence reads very odd to me "It will error and not apply any of the requested, valid updates should the net result of the patch result in an invalid state."

Please reword.

File: db/collection.go
Comment (shahzadlone) [inline db/collection.go:22]: question: I forgot if we cared about camel casing package import alias names. If we do then `jsonPatch`?

File: db/collection.go
Comment (shahzadlone) [inline db/collection.go:532]: suggestion: I would much rather than patch stuff be moved to a new file, like: `db/collection_patch.go`

File: db/collection.go
Comment (shahzadlone) [inline db/collection.go:684]: question: Why is this temporary? Will the use of reflection package be removed?

File: db/collection.go
Comment (shahzadlone) [inline db/collection.go:797]: question: `moves it` moves what?

File: tests/integration/collection_description/updates/add/collections_test.go
Comment (shahzadlone) [inline tests/integration/collection_description/updates/add/collections_test.go:33]: Would add another test where it is explicitly `0`:
`{ "op": "add", "path": "/2", "value": {"ID": 0, "Name": "Dogs"} }`

File: tests/integration/test_case.go
Comment (shahzadlone) [inline tests/integration/test_case.go:106]: todo: document please

File: cli/collection_patch.go
Comment (AndrewSisley) [reply cli/collection_patch.go:1]: :grin: Will do

- [ ] 2024

File: cli/collection_patch.go
Comment (AndrewSisley) [reply cli/collection_patch.go:58]: I don't really care :)  But I've added a param limit, arg length is now 0 or 1.

File: client/db.go
Comment (AndrewSisley) [reply client/db.go:128]: A valid CollectionDescription

File: client/db.go
Comment (AndrewSisley) [reply client/db.go:127]: What is odd about it?  Do you have a rough alternative in mind?

I think it is important to highlight that the full patch will be rolled back if an given patch item fails - especially given that SQLs do not do this and require manual rollback when given multiple DDL statements.

File: db/collection.go
Comment (AndrewSisley) [reply db/collection.go:684]: The validation (function) is temporary, which of course means that if this function is removed, the call to DeepEqual will be removed with it.

File: db/collection.go
Comment (AndrewSisley) [reply db/collection.go:797]: The collection relative to it's index in the map, I'll replace `it`

- [ ] Replace `it`

File: tests/integration/collection_description/updates/add/collections_test.go
Comment (AndrewSisley) [reply tests/integration/collection_description/updates/add/collections_test.go:33]: Will add

- [ ] Add explicit zero add test

File: tests/integration/test_case.go
Comment (AndrewSisley) [reply tests/integration/test_case.go:106]: My bad :)  Of course :)

- [ ] Doc test action Patch prop

File: db/collection.go
Comment (AndrewSisley) [reply db/collection.go:22]: I don't know either, my IDE did this, and I don't see it as being unreadable

File: db/collection.go
Comment (AndrewSisley) [reply db/collection.go:532]: I've opened a ticket: https://github.com/sourcenetwork/defradb/issues/2407

**Review by nasdf — CHANGES_REQUESTED** [review]
One minor fix from me.

File: client/db.go
Comment (shahzadlone) [reply client/db.go:127]: This sounds more clear when you said: `full patch will be rolled back if an given patch item fails` 

This is very wordy: `... apply any of the requested, valid updates should the net resulted`

File: db/collection.go
Comment (shahzadlone) [reply db/collection.go:684]: > The validation (function) is temporary, which means that if this function is removed, the call to DeepEqual will be removed with it.

Why is it temporary?

File: client/db.go
Comment (AndrewSisley) [reply client/db.go:127]: `rolled back` is incorrect/lazy lol - the patches are aggregated before being applied - hence the use of `net result of the patch`

File: tests/integration/collection_description/updates/replace/id_test.go
Comment (fredcarle) [inline tests/integration/collection_description/updates/replace/id_test.go:90]: thought: The error is a little weird here. It's talking about mutating sources when the patch is dealing with the collection ID.

**Review by fredcarle — COMMENTED** [review]
Just a few comments to resolve before approval :)

File: client/db.go
Comment (shahzadlone) [reply client/db.go:127]: > the patches are aggregated before being applied

This will be nice to document somewhere too

File: cli/collection_patch.go
Comment (shahzadlone) [reply cli/collection_patch.go:67]: Don't all other file consumption commands use `--file |-f` why do we want to have a different flag name for this arg (does it clash with `-f` somewhere?)

**Review by shahzadlone — APPROVED** [review]
LGTM assuming you will resolve other comments

File: db/collection.go
Comment (AndrewSisley) [reply db/collection.go:684]: Because the function is validating that indexes have not been modified, and we very much want them to be mutable in the future.

File: client/db.go
Comment (AndrewSisley) [reply client/db.go:127]: > I though I understood the bahaviour from the prior sentence but this one gets me confused. How can the net result be valid if an individual operation creates an invalid state?

There are many reasons why we want this to be, and many of the tests cover this, for example:

> { "op": "copy", "from": "/Users/Fields/1", "path": "/Users/Fields/2" },
{ "op": "replace", "path": "/Users/Fields/2/Name", "value": "age" },

The first field is copied (invalid as it is now a duplicate), and then renamed (net is now valid).  This allows existing stuff (including full schema/collections) to be used as templates.  There are other use cases too.

> This will be nice to document somewhere too

It is, we are talking about the documentation that currently documents it.

File: tests/integration/collection_description/updates/replace/id_test.go
Comment (AndrewSisley) [reply tests/integration/collection_description/updates/replace/id_test.go:90]: Yes, unfortunately when mutating IDs it becomes impossible for us to tell what the original object was (and that the ID was mutated), so it will often get caught by other validation rules.

File: db/collection.go
Comment (shahzadlone) [reply db/collection.go:684]: Thanks that makes so much sense.

Please document this

File: tests/integration/collection_description/updates/copy/name_test.go
Comment (AndrewSisley) [reply tests/integration/collection_description/updates/copy/name_test.go:43]: I think the value is low, as duplicate names are tested elsewhere, but I will add

- [ ] add copy name test

File: db/collection.go
Comment (AndrewSisley) [reply db/collection.go:797]: > { "op": "move", "from": "/1", "path": "/2" }

This moves the collection within the map, without mutating the ID.

File: cli/collection_patch.go
Comment (AndrewSisley) [reply cli/collection_patch.go:67]: `PatchSchema` uses `p` because it it takes two kinds of files.  I think it is more important to be consistent with `PatchSchema` than anywhere else here.

File: db/collection.go
Comment (fredcarle) [reply db/collection.go:797]: Thanks for clarifying.

File: cli/collection_patch.go
Comment (shahzadlone) [reply cli/collection_patch.go:67]: How does it take two files? I see 1 so far with -p

Comment (AndrewSisley) [reply cli/collection_patch.go:67]: `PatchSchema` takes two files, I assume you are looking at `PatchCollection`

Comment (fredcarle) [reply cli/collection_patch.go:67]: Maybe we should change the flag for patch schema to also be -f and make it consistent throughout. The lens file can stay with -t or maybe -l would make more sense.

Comment (AndrewSisley) [reply cli/collection_patch.go:67]: `f` and `file` is ambiguous for a command that takes two different files.  I do not like that.

Comment (fredcarle) [reply cli/collection_patch.go:67]: @nasdf your opinion would be appreciated here when you have a chance.

**Review by nasdf — APPROVED** [review]

File: cli/collection_patch.go
Comment (nasdf) [reply cli/collection_patch.go:67]: I think its fine to keep it `-p` for consistency with the other patch method. Using `-f` when there are multiple files is worse in my opinion.

**Pull request merged**
